### ISSUE TITLE
[BUG] False positive FLS when using custom access utility class

### ISSUE THREAD

**Issue by mfrenchlitify** [forcedotcom/sfdx-scanner#862]
**Describe the bug**
When using an internal utility class to check FLS and CRUD permissions, we sometimes get a false positive FLS item on the scan report. 

**To Reproduce**
Sample code that produces a false positive
```
    private static final Set<SObjectField> FIELDS = new Set<SObjectField>{
        Custom_Obj__c.Field_1__c,
        Custom_Obj__c.Field_2__c
    };

    //..........

    @AuraEnabled
    public static void insertCustomObj(String logString,String status){
        if(logString != null && logString != ''){ 
            List<Custom_Obj__c> ieList = new List<Custom_Obj__c>();
            if( AccessCheckUtility.isCreatable(Schema.sObjectType.Custom_Obj__c, FIELDS)){//access checked here
                    for(InvoiceImporterLogWrapper logWrapper:logList){
                        Custom_Obj__c ie = new Custom_Obj__c ();
                        ie.Field_1__c = somevalue.log;
                        ie.Field_2__c = somevalue.log
                        ieList.add(ie);
                    }
            }
            //Access checked above
            database.insert(ieList,false);//FLS violation here, despite access being checked above via utility class
        }    
    }


//AccessCheckUtility

    public static Boolean isCreatable(DescribeSObjectResult sObjectDescribe, Set<SObjectField> sObjectFieldSet) {
        if (sObjectDescribe == null || ! sObjectDescribe.isCreateable()) {
            String accessCreateableError = "Some Error";
            throw new AuraHandledException(accessCreateableError + sObjectDescribe.name + '.');
        }

        for (SObjectField sObjectField : sObjectFieldSet) {
            DescribeFieldResult describeFieldResult = DescribeCacheUtility.getFieldDescribe(sObjectDescribe.SObjectType, sObjectField);
            if (!describeFieldResult.isCreateable()) {
                String accessCreateableError = "Some Error"
                System.debug(System.LoggingLevel.ERROR, accessCreateableError + sObjectDescribe.name + '.' + sObjectField.getDescribe().label + ' field.');
                throw new AuraHandledException(accessCreateableError + sObjectDescribe.name + '.' + sObjectField.getDescribe().label + ' field.');
            }
        }

        return true;
    }
```

**Expected behavior**
No positive FLS or CRUD results for the above code

**Screenshots**

**Desktop (please complete the following information):**
 - OS: iOS
 - Scanner Version sfdx-cli/7.174.0 darwin-x64 node-v16.17.1

**Comment by rmohan20** [comment]
Hi @mfrenchlitify - thanks for letting us know.

I'm trying to reproduce this locally with your example code. Can you please help me with more information?

1. I'm assuming this is from the `scanner:run:dfa` command. Can you share the full violation message that was thrown? Feel free to mask the object name / field names to match your mockup.

2. Is `DescribeCacheUtility` a class in your project repo or an external dependency? Could share the content in `DescribeCacheUtility.getFieldDescribe(SObjectType type, SObjectField field)` or tell us more about it?

3. The `database.insert` in your sample is executed on `ieList`, but I can't find a definition within the scope of the code you've provided. I took the liberty to add a simplified definition - [see attached file](https://github.com/forcedotcom/sfdx-scanner/files/9892381/NoCrudFlsIssueRepro.cls.txt). Does this match your situation?

**Comment by mfrenchlitify** [comment]
1) Here is the message we are getting:
```
FLS validation is missing for [INSERT] operation on [Custom_Obj__c] with field(s) [Field_1__c,Field_2__c].
```

2) Here is the relevant contents of DescribeCacheUtility:
```
    public static Map<String, Schema.SObjectType> getGlobalDescribe() {
        if (globalDescribeStorage == null) {
            globalDescribeStorage = Schema.getGlobalDescribe();
        }
        return globalDescribeStorage;
    }

    public static DescribeSObjectResult getSObjectDescribe(SObjectType sObjectType) {
        String objectApiName = String.valueOf(sObjectType);
        return getSObjectDescribe(objectApiName);
    }

    public static DescribeSObjectResult getSObjectDescribe(String objectApiName) {
        if (String.isBlank(objectApiName)) {
            throw new DescribeCacheException(BLANK_OBJECT_API_NAME_EXCEPTION);
        }

        if (!sObjectsDescribeStorage.containsKey(objectApiName)) {
            SObjectType sObjectType = getGlobalDescribe().get(objectApiName);
            if (sObjectType == null) {
                throw new DescribeCacheException(UNKNOWN_OBJECT_API_NAME_EXCEPTION);
            }

            DescribeSObjectResult describeSObjectResult = sObjectType.getDescribe();
            sObjectsDescribeStorage.put(objectApiName, describeSObjectResult);
        }

        return sObjectsDescribeStorage.get(objectApiName);
    }

    public static DescribeFieldResult getFieldDescribe(SObjectType sObjectType, SObjectField sObjectField) {
        String objectApiName = String.valueOf(sObjectType);
        String fieldApiName = String.valueOf(sObjectField);
        return getFieldDescribe(objectApiName, fieldApiName);
    }

    public static DescribeFieldResult getFieldDescribe(String objectApiName, String fieldApiName) {
        Map<String, DescribeFieldResult> describeFieldResultsByApiName;

        if (String.isBlank(objectApiName) && String.isBlank(fieldApiName)) {
            throw new DescribeCacheException(BLANK_OBJECT_OR_FIELD_API_NAME_EXCEPTION);
        }

        if (!fieldsDescribeStorage.containsKey(objectApiName)) {
            describeFieldResultsByApiName = new Map<String, Schema.DescribeFieldResult>();
            fieldsDescribeStorage.put(objectApiName, describeFieldResultsByApiName);
        }

        if (!fieldsDescribeStorage.get(objectApiName).containsKey(fieldApiName)) {
            SObjectField sObjectField = getSObjectDescribe(objectApiName).fields.getMap().get(fieldApiName);
            if (sObjectField == null) {
                throw new DescribeCacheException(UNKNOWN_FIELD_API_NAME_EXCEPTION);
            }

            describeFieldResultsByApiName = fieldsDescribeStorage.get(objectApiName);
            describeFieldResultsByApiName.put(fieldApiName, sObjectField.getDescribe());
            fieldsDescribeStorage.put(objectApiName, describeFieldResultsByApiName);
        }

        return fieldsDescribeStorage.get(objectApiName).get(fieldApiName);
    }
```

3) my mistake, I must have removed the ieList variable while sanitizing the code. I've updated the comment above to make it more accurate.

**Comment by rmohan20** [comment]
@mfrenchlitify Thanks for the additional information and snippets.

I had to add this to your sample of DescribeCacheUtility to avoid internal errors:
```
	static Map<String, Schema.SObjectType> globalDescribeStorage = new Map<String, Schema.SObjectType>();
	static Map<String, DescribeSObjectResult> sObjectsDescribeStorage = new Map<String, DescribeSObjectResult>();
	static Map<String, Map<String, Schema.DescribeFieldResult>> fieldsDescribeStorage = new Map<String, Map<String, Schema.DescribeFieldResult>>();
```

Once I added the definitions, I get:
```
Executed engines: sfge. No rule violations found.
```

Attaching my version of the [DescribeCacheUtility.cls](https://github.com/forcedotcom/sfdx-scanner/files/9905740/DescribeCacheUtility.cls.txt) class for your reference.

Interestingly, when I remove `DescribeCacheUtility.cls` from the `--projectdir` directory, I get the same violation you've pasted:
```
FLS validation is missing for [INSERT] operation on [Custom_Obj__c] with field(s) [Field_1__c,Field_2__c].
```

My follow-up questions for you:

1. Do the shapes of variables (`globalDescribeStorage`, `sObjectsDescribeStorage`, `fieldsDescribeStorage`) I've pasted above match the ones you have in your code? If not, could you send me your definitions?
2. Can you confirm that your `DescribeCacheUtility.cls` file lives under the `--projectdir` location in your command? If it is recursively under, could you send me the structure/levels?

**Comment by rmohan20** [comment]
I spoke to Amit Tripathi today and he shared with me his version of the mockup that could repro this issue. This is a good lead to get closer to what's going on. Will post an update when I have one.

**Comment by mfrenchlitify** [comment]
@rmohan20 

Following up,

1) Here are the variable definitions, I've also added them to the code above. 
```
    @TestVisible
    private static Map<String, Schema.SObjectType> globalDescribeStorage;
    @TestVisible
    private static Map<String, DescribeSObjectResult> sObjectsDescribeStorage = new Map<String, DescribeSObjectResult>();
    @TestVisible
    private static Map<String, Map<String, DescribeFieldResult>> fieldsDescribeStorage = new Map<String, Map<String, DescribeFieldResult>>();
```
The variables you added were indeed correct. 

2) Yes, the `DescribeCacheUtility` class lives under the directory specified in our run command. Specifically, it lives at
`force-app/main/default/classes/DescribeCacheUtility.cls`, along with the rest of the apex classes.

**Comment by rmohan20** [comment]
@mfrenchlitify Thanks for confirming!

**Comment by rmohan20** [comment]
@mfrenchlitify I suddenly realized that in the snippet you've pasted, the actual `database.insert()` operation is outside the bounds of the access check, which means the database operation would happen irrespective of the outcome of the access check. This looks like a valid CRUD/FLS violation indeed - isn't it?

I'm still investigating why my interpretation of your code snippet did not throw a violation as well. Will report here when I have more information.

**Comment by rmohan20** [comment]
Fyi, we've figured out the root cause and are working on a solution. Will post another update once we know when it'll be available.
Thanks again for reporting this issue and helping us repro it.

**Comment by rmohan20** [comment]
@mfrenchlitify #893 should partially fix this issue, but another half related to handling for-loops correctly is still in progress. Will link this issue to both PRs once the changes are ready.
We'll be releasing the changes sometime next week.

**Comment by rmohan20** [comment]
@mfrenchlitify We've fixed this false positive with #923 and the fix should be available to you sometime this week. I'm closing this issue, but please do let us know if you notice something we should look into.
